no more magic

dscottboggs opened the pull request:
This commit also includes benchmarks proving the viability of reloading the magic database for every filetype request, should that become necessary.

Tested using the upload_photo example both with and without the `"magic"` feature.

dscottboggs committed:
Wrap magic cookie in a mutex

This commit also includes benchmarks proving the viability of reloading the
magic database for every filetype request, should that become necessary.

dscottboggs commented:
This will resolve #11

dscottboggs commented:
@ckruse and @matthiasbeyer

this shouldn't have any breaking changes with v1.0.1, seems ok to release as v1.1.0 or v1.0.3?

matthiasbeyer commented:
Maybe add some static assertions as well, to make sure that the type is actually `Send` as expected (and to prevent breaking it again :laughing: )

dscottboggs commented:
Oh? How would I do that?

matthiasbeyer commented:
The easy way: Pull `magic-is-not-send/patches` to this branch :wink: 

```
git pull --ff-only https://github.com/matthiasbeyer/mastodon-async magic-is-not-send/patches
```

or even better:

```
git fetch https://github.com/matthiasbeyer/mastodon-async magic-is-not-send/patches
git cherry-pick -x FETCH_HEAD
```

:wink:

dscottboggs commented:
No, this won't do. Even with the change adding it to a Mutex, the type is still poisoned by `!Send`. I'll keep the static assertion, but make the mime-type fetch a free function that reloads the database every time, instead.

dscottboggs committed:
Drop the magic bullshit

There were too many false positive responses from libmagic to have this
just happen in the background. In a future release I may add a specific
error which is returned when the given file has no extension, and
downstream application developers can use that as an indicator that they
should ask the user if they want to detect the filetype, making use of
libmagic directly in the downstream application as appropriate.

dscottboggs commented:
There were too many false positive responses from libmagic to have this
just happen in the background. ~~In a future release I may add a specific
error which is returned when the given file has no extension, and
downstream application developers can use that as an indicator that they
should ask the user if they want to detect the filetype, making use of
libmagic directly in the downstream application as appropriate.~~

I actually think that error would be better handled by the server. Sorry about this whole rabbit-hole. v1.0.3 will be out as soon as the CI clears and drops all this bullshit.

ckruse commented:
Great news! Thanks a lot!
